File: Sources/ObjC/BLECDevice.m
//
//  BLECDevice.m
//  Pods
//
//  Created by Balázs Kilvády on 6/1/16.
//
//

#import "BLECDevice.h"


@implementation BLECDeviceData

@end


@implementation BLECDevice

- (instancetype)initWithUUID:(NSUUID *)uuid
{
    self = [super init];
    if (self) {
        _UUID = uuid;
        _peripheral = nil;
        _characteristics = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (instancetype)initWithPeripheral:(CBPeripheral *)peripheral
{
    self = [super init];
    if (self) {
        _UUID = peripheral.identifier;
        _peripheral = peripheral;
        _characteristics = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (CBCharacteristic *)characteristicAt:(NSUInteger)characteristicIndex
                           inServiceAt:(NSUInteger)serviceIndex
{
    __block CBCharacteristic *found = nil;
    [_characteristics enumerateKeysAndObjectsUsingBlock:^(CBUUID *key, BLECDeviceData *data, BOOL *stop) {
        if (data.serviceIndex == serviceIndex && data.characteristicIndex == characteristicIndex) {
            found = data.characteristic;
            *stop = YES;
        }
    }];
    return found;
}

- (void)readRSSI
{
    if (_peripheral) {
        [_peripheral readRSSI];
    }
}

- (void)writeValue:(NSData *)value
 forCharacteristic:(CBCharacteristic *)characteristic
      WithResponse:(void (^)(NSError *error))response
{
    if (!_peripheral) return;

    if (response == nil) {
        [_peripheral writeValue:value
              forCharacteristic:characteristic
                           type:CBCharacteristicWriteWithoutResponse];
    } else {
        BLECDeviceData *charData = _characteristics[characteristic.UUID];
        if (charData == nil) {
            return;
        }
        if (charData.writeResponse != nil) {
            NSAssert(NO, @"UNREACHABLE");
            return;
        }
        charData.writeResponse = response;
        _characteristics[characteristic.UUID] = charData;
        [_peripheral writeValue:value
              forCharacteristic:characteristic
                           type:CBCharacteristicWriteWithResponse];
    }
}

@end

File: Sources/Swift/BLECDevice.swift
//
//  BLECDevice.swift
//  Pods
//
//  Created by Balázs Kilvády on 7/4/16.
//
//

import Foundation
import CoreBluetooth


struct BLECPeripheralState: OptionSet {
    let rawValue: UInt

    static let None       = BLECPeripheralState(rawValue: 0)
    static let Discovered = BLECPeripheralState(rawValue: 0x01)
    static let Connected  = BLECPeripheralState(rawValue: 0x02)

    init(rawValue: UInt) {
        self.rawValue = rawValue
    }
}

struct BLECDeviceData {
    var characteristic: CBCharacteristic
    var delegate: BLECCharacteristicDelegate
    var serviceIndex: Int
    var characteristicIndex: Int
    var writeResponse: ((Error?) -> Void)?
}

open class BLECDevice {

    public enum DeviceError: Error {
        case noPeripheral
        case invalidCharacteristic
        case alredyPending
    }

    open var peripheral: CBPeripheral?
    var state = BLECPeripheralState.None
    open let UUID: Foundation.UUID
    var characteristics = [CBUUID: BLECDeviceData]()

    init(UUID uuid: Foundation.UUID) {
        UUID = uuid
        peripheral = nil
    }

    init(peripheral: CBPeripheral) {
        UUID = peripheral.identifier
        self.peripheral = peripheral
    }

    open func characteristicAt(_ characteristicIndex: Int, inServiceAt serviceIndex: Int) -> CBCharacteristic? {
        for (_, data) in characteristics {
            if data.serviceIndex == serviceIndex && data.characteristicIndex == characteristicIndex {
                return data.characteristic
            }
        }
        return nil
    }

    open func readRSSI() throws {
        guard let peripheral = self.peripheral else {
            throw DeviceError.noPeripheral
        }

        peripheral.readRSSI()
    }

    open func writeValue(_ data: Data,
                           forCharacteristic characteristic: CBCharacteristic,
                           response: ((Error?) -> Void)?) throws {
        guard let peripheral = self.peripheral else {
            throw DeviceError.noPeripheral
        }

        if response == nil {
            peripheral.writeValue(data, for: characteristic, type: .withoutResponse)
        } else {
            guard var charData = self.characteristics[characteristic.uuid] else {
                throw DeviceError.invalidCharacteristic
            }
            if charData.writeResponse != nil {
                throw DeviceError.alredyPending
            }
            charData.writeResponse = response!
            self.characteristics[characteristic.uuid] = charData
            peripheral.writeValue(data, for: characteristic, type: .withResponse)
        }
    }
}
